docs/proposals/query-blocking.md

```diff
@@ -0,0 +1,101 @@
+
+**Author:** @wilfriedroset
+
+**Date:** 05/2023
+
+**Sponsor(s):** @username of maintainer(s) willing to shepherd this MID
+
+**Type:** Feature
+
+**Status:** Draft
+
+**Related issues/PRs:**
+
+---
+
+## Background
+
+The `query-frontend` is a component that is responsible for query acceleration through splitting, sharding, caching, and other techniques. It is also responsible for enforcing limits on queries such as total query time range.
+
+## Problem Statement
+
+In certain situations, you may not be able to control the queries being sent to your Mimir installation. These queries may be intentionally or unintentionally expensive to run, and they may affect the overall stability or cost of running your service.
+
+Example:
+
+```
+sum(increase(tfe_run_current_count{status=~"applied|planned_and_finished", instance="host1", organization_name=~".*", workspace_name=~".*"}[2674800s]) > 0)
+```
+
+The query above can be harmful as the wildcard regex would match any label values including empty string. The result of such query is not cached by Mimir. As such the read path would be under pressure if the query is repeated by the tenant (e.g: dashboard in auto repeat).
+
+## Goals
+
+- Prevent harmful queries to be processed by Mimir's read path.
+
+## Non-Goals (optional)
+
+- Rewrite queries in a non harmful way.
+- Optimize queries on the fly.
+- Blocked remote read.
+- Block similar queries with a single definition. Blocking `metric_name{label1=value1, label2=value2}` does not automatically block `metric_name{label2=value2, label1=value1}`.
+- Lint or re-format queries.
+
+## Proposals
+
+### Proposal 0: Do nothing
+
+Current tenants isolation through `shuffle-sharding` are sufficient to keep the read path healthy. Mimir's operators can use limits such as `max_fetched_chunks_per_query`, `max_queriers_per_tenant` or `store_gateway_tenant_shard_size` to limits tenants processing capabilities.
+
+**Pros:**
+
+- Nothing to be done
+
+**Cons:**
+
+- Harmful queries are still processed by Mimir but in a limited part of the cluster
+
+### Proposal 1: Query blocking
+
+Based on the configuration, Mimir's [limitsMiddleware](https://github.com/grafana/mimir/blob/main/pkg/frontend/querymiddleware/limits.go)) runned by the `query-frontend` would be responsible for blocking queries. Operators would be able to block queries either for all tenants or on a per-tenant basis via the _runtime configuration_.
+
+```yaml
+overrides:
+  "tenant-id":
+    blocked_queries:
+      # block this query exactly
+      - pattern: 'sum(rate({env="prod"}[1m]))'
+
+      # block any query matching this regex pattern
+      - pattern: '="\.\*"' # all queries with a label value `.*`
+        regex: true
+```
+
+The order of patterns is preserved, so the first matching pattern will be used. If `blocked_queries` are defined in both _runtime_configuration_ and _configuration_, the resulting list should be the concatenation of both with the _runtime_configuration_ first.
```

Comment:
I'm not sure about the "concatenation". Currently the runtime configuration, where you specify the per-tenant limits simply takes precedence over the standard one. I would start keeping this behaviour.